apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/dynamic/ConfigurationDiscoveryService.java

```diff
@@ -121,7 +121,8 @@ public void shutdown() throws Throwable {
     public void registerAgentConfigChangeWatcher(AgentConfigChangeWatcher watcher) {
         WatcherHolder holder = new WatcherHolder(watcher);
         if (register.containsKey(holder.getKey())) {
-            throw new IllegalStateException("Duplicate register, watcher=" + watcher);
+            LOGGER.debug("Duplicate register, watcher={}", watcher);
+            return;
```

Comment:
I think we need to enhance this logic. Duplicate watchers should not be always legal and ignore the second. We should check the input class and existing watcher class (including class loader level), if they are the same, we should be able to ignore the duplicate.
Think in this way, one config class could be loaded by different Agent ClassLoader objects as it is being used in two classloaders of service's codes.
If they are not matched, we should not ignore them, we should support more than one watcher instance in the `ConfigurationDiscoveryService`.

In your current implementation, the dynamic config updating could work on one class loader's SQL parameter collecting, but not for others, and randomly. This would be a very confused bug, and hard to locate when others face this.

Reply:
> Think in this way, one config class could be loaded by different Agent ClassLoader objects as it is being used in two classloaders of service's codes.

But most time, we just register watcher in one place, so do not exist watcher is loaded by two classloader. Is it cost-effective to fit the situation we rarely use?

Reply:
Remember, you are changing agent core, so, some extra list is definitely acceptable as they are only created once.
But making core having known bugs seems not a good practice. After all, multiple class loaders are common in some runtime env.

Reply:
I was writing this kernel with much smaller cases, if I said users will not face those cases, people like you will be in a chaos.

Reply:
Ok, can I change `ConfigurationDiscoveryService.Register#register` field from `Map<String, WatcherHolder> register` to `Map<String, List<WatcherHolder>> register` for supporting multi same key watcher?

Reply:
Yes, you should follow this way.